File: scripts/migration/neo4j/migrate.go
// Neo4j → NornicDB migration (Go).
//
// Bolt → Bolt. Reads constraints, indexes, nodes, and relationships from a
// source Neo4j instance and replays them into NornicDB using hot-path
// UNWIND/MERGE shapes from docs/performance/hot-path-query-cookbook.md.
//
// Phases (always in this order):
//
//  1. Schema  — constraints, then indexes (vector and fulltext last).
//  2. Nodes   — UNWIND $rows MERGE on _neo4j_id key, SET props.
//                Hits UnwindSimpleMergeBatch.
//  3. Edges   — UNWIND $rows MATCH (a), MATCH (b) CREATE edge.
//                Hits UnwindMultiMatchCreateBatch.
//
// Run:
//
//   go run ./scripts/migration/neo4j/migrate.go \
//       --source-url bolt://neo4j.prod:7687 --source-user neo4j --source-pass <pw> \
//       --target-url bolt://nornicdb.local:7687 \
//       --batch-size 500
//
// Flags: --skip-schema, --skip-nodes, --skip-edges, --labels A,B, --types T,U,
// --dry-run.

package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

type config struct {
	sourceURL, sourceUser, sourcePass, sourceDB string
	targetURL, targetUser, targetPass, targetDB string
	batchSize                                   int
	skipSchema, skipNodes, skipEdges            bool
	labels, types                               string
	dryRun                                      bool
}

func parseFlags() config {
	cfg := config{}
	flag.StringVar(&cfg.sourceURL, "source-url", "", "bolt://host:7687")
	flag.StringVar(&cfg.sourceUser, "source-user", "neo4j", "")
	flag.StringVar(&cfg.sourcePass, "source-pass", "", "")
	flag.StringVar(&cfg.sourceDB, "source-database", "neo4j", "")
	flag.StringVar(&cfg.targetURL, "target-url", "bolt://localhost:7687", "")
	flag.StringVar(&cfg.targetUser, "target-user", "neo4j", "")
	flag.StringVar(&cfg.targetPass, "target-pass", "password", "")
	flag.StringVar(&cfg.targetDB, "target-database", "neo4j", "")
	flag.IntVar(&cfg.batchSize, "batch-size", 500, "rows per UNWIND batch")
	flag.BoolVar(&cfg.skipSchema, "skip-schema", false, "")
	flag.BoolVar(&cfg.skipNodes, "skip-nodes", false, "")
	flag.BoolVar(&cfg.skipEdges, "skip-edges", false, "")
	flag.StringVar(&cfg.labels, "labels", "", "comma-separated label filter")
	flag.StringVar(&cfg.types, "types", "", "comma-separated relationship-type filter")
	flag.BoolVar(&cfg.dryRun, "dry-run", false, "")
	flag.Parse()
	if cfg.sourceURL == "" || cfg.sourcePass == "" {
		log.Fatal("--source-url and --source-pass are required")
	}
	return cfg
}

func csvSet(s string) map[string]struct{} {
	out := map[string]struct{}{}
	for _, x := range strings.Split(s, ",") {
		if t := strings.TrimSpace(x); t != "" {
			out[t] = struct{}{}
		}
	}
	return out
}

// ---- Schema ----------------------------------------------------------------

func fetchRecords(ctx context.Context, sess neo4j.SessionWithContext, q string) ([]map[string]any, error) {
	res, err := sess.Run(ctx, q, nil)
	if err != nil {
		return nil, err
	}
	out := []map[string]any{}
	for res.Next(ctx) {
		out = append(out, res.Record().AsMap())
	}
	return out, res.Err()
}

func toStringSlice(v any) []string {
	xs, ok := v.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(xs))
	for _, x := range xs {
		if s, ok := x.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

func constraintToCypher(c map[string]any) string {
	name, _ := c["name"].(string)
	ctype := strings.ToUpper(fmt.Sprint(c["type"]))
	entity := strings.ToUpper(fmt.Sprint(c["entityType"]))
	labels := toStringSlice(c["labelsOrTypes"])
	props := toStringSlice(c["properties"])
	if len(labels) == 0 || len(props) == 0 {
		return ""
	}
	label := labels[0]

	propParts := make([]string, len(props))
	for i, p := range props {
		propParts[i] = "n." + p
	}
	propList := strings.Join(propParts, ", ")
	if len(props) > 1 {
		propList = "(" + propList + ")"
	}

	switch entity {
	case "NODE":
		switch ctype {
		case "UNIQUENESS":
			return fmt.Sprintf("CREATE CONSTRAINT %s IF NOT EXISTS FOR (n:%s) REQUIRE %s IS UNIQUE", name, label, propList)
		case "NODE_KEY":
			return fmt.Sprintf("CREATE CONSTRAINT %s IF NOT EXISTS FOR (n:%s) REQUIRE %s IS NODE KEY", name, label, propList)
		case "NODE_PROPERTY_EXISTENCE":
			return fmt.Sprintf("CREATE CONSTRAINT %s IF NOT EXISTS FOR (n:%s) REQUIRE %s IS NOT NULL", name, label, propList)
		}
	case "RELATIONSHIP":
		switch ctype {
		case "RELATIONSHIP_PROPERTY_EXISTENCE":
			return fmt.Sprintf("CREATE CONSTRAINT %s IF NOT EXISTS FOR ()-[r:%s]-() REQUIRE r.%s IS NOT NULL", name, label, props[0])
		case "RELATIONSHIP_UNIQUENESS":
			return fmt.Sprintf("CREATE CONSTRAINT %s IF NOT EXISTS FOR ()-[r:%s]-() REQUIRE r.%s IS UNIQUE", name, label, props[0])
		}
	}
	return ""
}

func indexToCypher(idx map[string]any) string {
	if oc, _ := idx["owningConstraint"].(string); oc != "" {
		return ""
	}
	name, _ := idx["name"].(string)
	itype := strings.ToUpper(fmt.Sprint(idx["type"]))
	entity := strings.ToUpper(fmt.Sprint(idx["entityType"]))
	labels := toStringSlice(idx["labelsOrTypes"])
	props := toStringSlice(idx["properties"])
	if len(labels) == 0 || len(props) == 0 || entity != "NODE" {
		return ""
	}
	label := labels[0]

	switch itype {
	case "RANGE", "BTREE":
		if len(props) == 1 {
			return fmt.Sprintf("CREATE INDEX %s IF NOT EXISTS FOR (n:%s) ON (n.%s)", name, label, props[0])
		}
		propParts := make([]string, len(props))
		for i, p := range props {
			propParts[i] = "n." + p
		}
		return fmt.Sprintf("CREATE INDEX %s IF NOT EXISTS FOR (n:%s) ON (%s)", name, label, strings.Join(propParts, ", "))
	case "FULLTEXT":
		propParts := make([]string, len(props))
		for i, p := range props {
			propParts[i] = "n." + p
		}
		return fmt.Sprintf("CREATE FULLTEXT INDEX %s IF NOT EXISTS FOR (n:%s) ON EACH [%s]", name, label, strings.Join(propParts, ", "))
	case "VECTOR":
		opts, _ := idx["options"].(map[string]any)
		cfg, _ := opts["indexConfig"].(map[string]any)
		dims, _ := cfg["vector.dimensions"].(int64)
		sim, _ := cfg["vector.similarity_function"].(string)
		if dims == 0 {
			return ""
		}
		if sim == "" {
			sim = "cosine"
		}
		return fmt.Sprintf(
			"CREATE VECTOR INDEX %s IF NOT EXISTS FOR (n:%s) ON (n.%s) "+
				"OPTIONS {indexConfig: {`vector.dimensions`: %d, `vector.similarity_function`: '%s'}}",
			name, label, props[0], dims, sim,
		)
	}
	return ""
}

func replicateSchema(ctx context.Context, src, dst neo4j.SessionWithContext, dryRun bool) error {
	fmt.Println("\n[schema]")
	cs, err := fetchRecords(ctx, src, "SHOW CONSTRAINTS")
	if err != nil {
		return fmt.Errorf("fetch constraints: %w", err)
	}
	for _, c := range cs {
		ddl := constraintToCypher(c)
		if ddl == "" {
			fmt.Printf("  · skip constraint (unmapped): %v\n", c["name"])
			continue
		}
		fmt.Printf("  → %s\n", ddl)
		if dryRun {
			continue
		}
		if _, err := dst.Run(ctx, ddl, nil); err != nil {
			return fmt.Errorf("apply constraint: %w", err)
		}
	}

	is, err := fetchRecords(ctx, src, "SHOW INDEXES")
	if err != nil {
		return fmt.Errorf("fetch indexes: %w", err)
	}
	for _, i := range is {
		ddl := indexToCypher(i)
		if ddl == "" {
			continue
		}
		fmt.Printf("  → %s\n", ddl)
		if dryRun {
			continue
		}
		if _, err := dst.Run(ctx, ddl, nil); err != nil {
			return fmt.Errorf("apply index: %w", err)
		}
	}
	return nil
}

// ---- Nodes -----------------------------------------------------------------

const nodeUpsertCypher = `
UNWIND $rows AS row
MERGE (n:_Migrated {_neo4j_id: row.id})
SET n += row.props
SET n._neo4j_labels = row.labels
`

func fetchNodeLabels(ctx context.Context, sess neo4j.SessionWithContext) ([]string, error) {
	rs, err := fetchRecords(ctx, sess, "CALL db.labels() YIELD label RETURN label")
	if err != nil {
		return nil, err
	}
	out := make([]string, 0, len(rs))
	for _, r := range rs {
		if s, ok := r["label"].(string); ok {
			out = append(out, s)
		}
	}
	return out, nil
}

func streamNodes(ctx context.Context, sess neo4j.SessionWithContext, label string, batch int, yield func([]map[string]any) error) error {
	var lastID any
	for {
		var (
			res neo4j.ResultWithContext
			err error
		)
		if lastID == nil {
			q := fmt.Sprintf(
				"MATCH (n:`%s`) "+
					"RETURN elementId(n) AS id, labels(n) AS labels, properties(n) AS props "+
					"ORDER BY elementId(n) LIMIT $limit",
				label,
			)
			res, err = sess.Run(ctx, q, map[string]any{"limit": batch})
		} else {
			q := fmt.Sprintf(
				"MATCH (n:`%s`) WHERE elementId(n) > $cursor "+
					"RETURN elementId(n) AS id, labels(n) AS labels, properties(n) AS props "+
					"ORDER BY elementId(n) LIMIT $limit",
				label,
			)
			res, err = sess.Run(ctx, q, map[string]any{"cursor": lastID, "limit": batch})
		}
		if err != nil {
			return err
		}
		var rows []map[string]any
		for res.Next(ctx) {
			rows = append(rows, res.Record().AsMap())
		}
		if err := res.Err(); err != nil {
			return err
		}
		if len(rows) == 0 {
			return nil
		}
		if err := yield(rows); err != nil {
			return err
		}
		lastID = rows[len(rows)-1]["id"]
		if len(rows) < batch {
			return nil
		}
	}
}

func replicateNodes(ctx context.Context, src, dst neo4j.SessionWithContext, labels []string, filter map[string]struct{}, batch int, dryRun bool) (int, error) {
	fmt.Println("\n[nodes]")
	total := 0
	for _, label := range labels {
		if len(filter) > 0 {
			if _, ok := filter[label]; !ok {
				continue
			}
		}
		if strings.HasPrefix(label, "_") {
			continue
		}
		moved := 0
		err := streamNodes(ctx, src, label, batch, func(chunk []map[string]any) error {
			if !dryRun {
				if _, err := dst.Run(ctx, nodeUpsertCypher, map[string]any{"rows": chunk}); err != nil {
					return err
				}
			}
			moved += len(chunk)
			total += len(chunk)
			fmt.Printf("  · :%s: %d\n", label, moved)
			return nil
		})
		if err != nil {
			return total, err
		}
	}
	fmt.Printf("  total nodes upserted: %d\n", total)
	return total, nil
}

// ---- Edges -----------------------------------------------------------------

func fetchRelTypes(ctx context.Context, sess neo4j.SessionWithContext) ([]string, error) {
	rs, err := fetchRecords(ctx, sess, "CALL db.relationshipTypes() YIELD relationshipType RETURN relationshipType")
	if err != nil {
		return nil, err
	}
	out := make([]string, 0, len(rs))
	for _, r := range rs {
		if s, ok := r["relationshipType"].(string); ok {
			out = append(out, s)
		}
	}
	return out, nil
}

func streamEdges(ctx context.Context, sess neo4j.SessionWithContext, relType string, batch int, yield func([]map[string]any) error) error {
	var lastID any
	for {
		var (
			res neo4j.ResultWithContext
			err error
		)
		if lastID == nil {
			q := fmt.Sprintf(
				"MATCH (a)-[r:`%s`]->(b) "+
					"RETURN elementId(r) AS id, elementId(a) AS startId, elementId(b) AS endId, properties(r) AS props "+
					"ORDER BY elementId(r) LIMIT $limit",
				relType,
			)
			res, err = sess.Run(ctx, q, map[string]any{"limit": batch})
		} else {
			q := fmt.Sprintf(
				"MATCH (a)-[r:`%s`]->(b) WHERE elementId(r) > $cursor "+
					"RETURN elementId(r) AS id, elementId(a) AS startId, elementId(b) AS endId, properties(r) AS props "+
					"ORDER BY elementId(r) LIMIT $limit",
				relType,
			)
			res, err = sess.Run(ctx, q, map[string]any{"cursor": lastID, "limit": batch})
		}
		if err != nil {
			return err
		}
		var rows []map[string]any
		for res.Next(ctx) {
			rows = append(rows, res.Record().AsMap())
		}
		if err := res.Err(); err != nil {
			return err
		}
		if len(rows) == 0 {
			return nil
		}
		if err := yield(rows); err != nil {
			return err
		}
		lastID = rows[len(rows)-1]["id"]
		if len(rows) < batch {
			return nil
		}
	}
}

func edgeCreateCypher(relType string) string {
	return fmt.Sprintf(
		"UNWIND $rows AS row "+
			"MATCH (a:_Migrated {_neo4j_id: row.startId}) "+
			"MATCH (b:_Migrated {_neo4j_id: row.endId}) "+
			"CREATE (a)-[:`%s` {_neo4j_id: row.id}]->(b)",
		relType,
	)
}

func edgePropsCypher(relType string) string {
	return fmt.Sprintf(
		"UNWIND $rows AS row "+
			"MATCH (a:_Migrated {_neo4j_id: row.startId})-[r:`%s` "+
			"{_neo4j_id: row.id}]->(b:_Migrated {_neo4j_id: row.endId}) "+
			"SET r += row.props",
		relType,
	)
}

func replicateEdges(ctx context.Context, src, dst neo4j.SessionWithContext, types []string, filter map[string]struct{}, batch int, dryRun bool) (int, error) {
	fmt.Println("\n[edges]")
	total := 0
	for _, t := range types {
		if len(filter) > 0 {
			if _, ok := filter[t]; !ok {
				continue
			}
		}
		moved := 0
		createQ := edgeCreateCypher(t)
		propsQ := edgePropsCypher(t)
		err := streamEdges(ctx, src, t, batch, func(chunk []map[string]any) error {
			if !dryRun {
				if _, err := dst.Run(ctx, createQ, map[string]any{"rows": chunk}); err != nil {
					return err
				}
				hasProps := false
				for _, r := range chunk {
					if p, ok := r["props"].(map[string]any); ok && len(p) > 0 {
						hasProps = true
						break
					}
				}
				if hasProps {
					if _, err := dst.Run(ctx, propsQ, map[string]any{"rows": chunk}); err != nil {
						return err
					}
				}
			}
			moved += len(chunk)
			total += len(chunk)
			fmt.Printf("  · :%s: %d\n", t, moved)
			return nil
		})
		if err != nil {
			return total, err
		}
	}
	fmt.Printf("  total edges created: %d\n", total)
	return total, nil
}

// ---- Main ------------------------------------------------------------------

func main() {
	cfg := parseFlags()
	labelFilter := csvSet(cfg.labels)
	typeFilter := csvSet(cfg.types)

	ctx, cancel := context.WithTimeout(context.Background(), 12*time.Hour)
	defer cancel()

	srcDrv, err := neo4j.NewDriverWithContext(cfg.sourceURL, neo4j.BasicAuth(cfg.sourceUser, cfg.sourcePass, ""))
	if err != nil {
		log.Fatalf("source driver: %v", err)
	}
	defer srcDrv.Close(ctx)
	dstDrv, err := neo4j.NewDriverWithContext(cfg.targetURL, neo4j.BasicAuth(cfg.targetUser, cfg.targetPass, ""))
	if err != nil {
		log.Fatalf("target driver: %v", err)
	}
	defer dstDrv.Close(ctx)

	src := srcDrv.NewSession(ctx, neo4j.SessionConfig{DatabaseName: cfg.sourceDB, AccessMode: neo4j.AccessModeRead})
	defer src.Close(ctx)
	dst := dstDrv.NewSession(ctx, neo4j.SessionConfig{DatabaseName: cfg.targetDB, AccessMode: neo4j.AccessModeWrite})
	defer dst.Close(ctx)

	started := time.Now()

	if !cfg.skipSchema {
		if err := replicateSchema(ctx, src, dst, cfg.dryRun); err != nil {
			log.Fatalf("schema: %v", err)
		}
	}
	if !cfg.skipNodes {
		labels, err := fetchNodeLabels(ctx, src)
		if err != nil {
			log.Fatalf("fetch labels: %v", err)
		}
		if _, err := replicateNodes(ctx, src, dst, labels, labelFilter, cfg.batchSize, cfg.dryRun); err != nil {
			log.Fatalf("nodes: %v", err)
		}
	}
	if !cfg.skipEdges {
		types, err := fetchRelTypes(ctx, src)
		if err != nil {
			log.Fatalf("fetch types: %v", err)
		}
		if _, err := replicateEdges(ctx, src, dst, types, typeFilter, cfg.batchSize, cfg.dryRun); err != nil {
			log.Fatalf("edges: %v", err)
		}
	}

	fmt.Printf("\nDone in %s\n", time.Since(started).Round(time.Second))
}

File: scripts/migration/neo4j/migrate.mjs
#!/usr/bin/env node
// Neo4j → NornicDB migration (Node).
//
// Bolt → Bolt. Reads constraints, indexes, nodes, and relationships from a
// source Neo4j instance and replays them into NornicDB using hot-path
// UNWIND/MERGE shapes from docs/performance/hot-path-query-cookbook.md.
//
// Phases:
//   1. Schema  — constraints, then indexes (vector and fulltext last).
//   2. Nodes   — UNWIND $rows MERGE on _neo4j_id. Hits UnwindSimpleMergeBatch.
//   3. Edges   — UNWIND $rows MATCH (a), MATCH (b) CREATE edge.
//                Hits UnwindMultiMatchCreateBatch.
//
// Install:
//   npm install neo4j-driver
//
// Run:
//   node migrate.mjs \
//     --source-url bolt://neo4j.prod:7687 --source-user neo4j --source-pass <pw> \
//     --target-url bolt://nornicdb.local:7687 \
//     --batch-size 500
//
// Flags: --skip-schema, --skip-nodes, --skip-edges, --labels A,B, --types T,U,
// --dry-run.

import neo4j from "neo4j-driver";

function parseArgs(argv) {
  const args = {
    sourceUrl: "",
    sourceUser: "neo4j",
    sourcePass: "",
    sourceDb: "neo4j",
    targetUrl: "bolt://localhost:7687",
    targetUser: "neo4j",
    targetPass: "password",
    targetDb: "neo4j",
    batchSize: 500,
    skipSchema: false,
    skipNodes: false,
    skipEdges: false,
    labels: "",
    types: "",
    dryRun: false,
  };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    const next = () => argv[++i];
    switch (a) {
      case "--source-url":      args.sourceUrl = next(); break;
      case "--source-user":     args.sourceUser = next(); break;
      case "--source-pass":     args.sourcePass = next(); break;
      case "--source-database": args.sourceDb = next(); break;
      case "--target-url":      args.targetUrl = next(); break;
      case "--target-user":     args.targetUser = next(); break;
      case "--target-pass":     args.targetPass = next(); break;
      case "--target-database": args.targetDb = next(); break;
      case "--batch-size":      args.batchSize = parseInt(next(), 10); break;
      case "--skip-schema":     args.skipSchema = true; break;
      case "--skip-nodes":      args.skipNodes = true; break;
      case "--skip-edges":      args.skipEdges = true; break;
      case "--labels":          args.labels = next(); break;
      case "--types":           args.types = next(); break;
      case "--dry-run":         args.dryRun = true; break;
      default:
        console.error(`unknown flag: ${a}`);
        process.exit(2);
    }
  }
  if (!args.sourceUrl || !args.sourcePass) {
    console.error("--source-url and --source-pass are required");
    process.exit(2);
  }
  return args;
}

const csvSet = (s) =>
  new Set(s.split(",").map((x) => x.trim()).filter(Boolean));

// ---- Schema ----------------------------------------------------------------

async function fetchAll(sess, q) {
  const r = await sess.run(q);
  return r.records.map((rec) => Object.fromEntries(rec.keys.map((k) => [k, rec.get(k)])));
}

function constraintToCypher(c) {
  const name = c.name ?? "";
  const ctype = String(c.type ?? "").toUpperCase();
  const entity = String(c.entityType ?? "").toUpperCase();
  const labels = c.labelsOrTypes ?? [];
  const props = c.properties ?? [];
  if (!labels.length || !props.length) return null;
  const label = labels[0];
  const propParts = props.map((p) => `n.${p}`);
  const propList = props.length > 1 ? `(${propParts.join(", ")})` : propParts[0];

  if (entity === "NODE") {
    if (ctype === "UNIQUENESS")
      return `CREATE CONSTRAINT ${name} IF NOT EXISTS FOR (n:${label}) REQUIRE ${propList} IS UNIQUE`;
    if (ctype === "NODE_KEY")
      return `CREATE CONSTRAINT ${name} IF NOT EXISTS FOR (n:${label}) REQUIRE ${propList} IS NODE KEY`;
    if (ctype === "NODE_PROPERTY_EXISTENCE")
      return `CREATE CONSTRAINT ${name} IF NOT EXISTS FOR (n:${label}) REQUIRE ${propList} IS NOT NULL`;
  } else if (entity === "RELATIONSHIP") {
    if (ctype === "RELATIONSHIP_PROPERTY_EXISTENCE")
      return `CREATE CONSTRAINT ${name} IF NOT EXISTS FOR ()-[r:${label}]-() REQUIRE r.${props[0]} IS NOT NULL`;
    if (ctype === "RELATIONSHIP_UNIQUENESS")
      return `CREATE CONSTRAINT ${name} IF NOT EXISTS FOR ()-[r:${label}]-() REQUIRE r.${props[0]} IS UNIQUE`;
  }
  return null;
}

function indexToCypher(idx) {
  if (idx.owningConstraint) return null;
  const name = idx.name ?? "";
  const itype = String(idx.type ?? "").toUpperCase();
  const entity = String(idx.entityType ?? "").toUpperCase();
  const labels = idx.labelsOrTypes ?? [];
  const props = idx.properties ?? [];
  if (!labels.length || !props.length || entity !== "NODE") return null;
  const label = labels[0];

  if (itype === "RANGE" || itype === "BTREE") {
    if (props.length === 1)
      return `CREATE INDEX ${name} IF NOT EXISTS FOR (n:${label}) ON (n.${props[0]})`;
    return `CREATE INDEX ${name} IF NOT EXISTS FOR (n:${label}) ON (${props
      .map((p) => `n.${p}`)
      .join(", ")})`;
  }
  if (itype === "FULLTEXT") {
    return `CREATE FULLTEXT INDEX ${name} IF NOT EXISTS FOR (n:${label}) ON EACH [${props
      .map((p) => `n.${p}`)
      .join(", ")}]`;
  }
  if (itype === "VECTOR") {
    const cfg = idx.options?.indexConfig ?? {};
    const dims = Number(cfg["vector.dimensions"]);
    const sim = cfg["vector.similarity_function"] ?? "cosine";
    if (!Number.isFinite(dims) || dims <= 0) return null;
    return `CREATE VECTOR INDEX ${name} IF NOT EXISTS FOR (n:${label}) ON (n.${props[0]}) OPTIONS {indexConfig: {\`vector.dimensions\`: ${dims}, \`vector.similarity_function\`: '${sim}'}}`;
  }
  return null;
}

async function replicateSchema(srcSess, dstSess, dryRun) {
  console.log("\n[schema]");
  const constraints = await fetchAll(srcSess, "SHOW CONSTRAINTS");
  for (const c of constraints) {
    const ddl = constraintToCypher(c);
    if (!ddl) {
      console.log(`  · skip constraint (unmapped): ${c.name}`);
      continue;
    }
    console.log(`  → ${ddl}`);
    if (!dryRun) await dstSess.run(ddl);
  }

  const indexes = await fetchAll(srcSess, "SHOW INDEXES");
  for (const i of indexes) {
    const ddl = indexToCypher(i);
    if (!ddl) continue;
    console.log(`  → ${ddl}`);
    if (!dryRun) await dstSess.run(ddl);
  }
}

// ---- Nodes ----------------------------------------------------------------

const NODE_UPSERT_CYPHER = `
UNWIND $rows AS row
MERGE (n:_Migrated {_neo4j_id: row.id})
SET n += row.props
SET n._neo4j_labels = row.labels
`;

async function fetchLabels(sess) {
  const rs = await fetchAll(sess, "CALL db.labels() YIELD label RETURN label");
  return rs.map((r) => r.label);
}

async function* streamNodes(sess, label, batch) {
  let lastId = null;
  for (;;) {
    let res;
    if (lastId === null) {
      res = await sess.run(
        `MATCH (n:\`${label}\`) RETURN elementId(n) AS id, labels(n) AS labels, properties(n) AS props ORDER BY elementId(n) LIMIT $limit`,
        { limit: neo4j.int(batch) },
      );
    } else {
      res = await sess.run(
        `MATCH (n:\`${label}\`) WHERE elementId(n) > $cursor RETURN elementId(n) AS id, labels(n) AS labels, properties(n) AS props ORDER BY elementId(n) LIMIT $limit`,
        { cursor: lastId, limit: neo4j.int(batch) },
      );
    }
    const rows = res.records.map((rec) => ({
      id: rec.get("id"),
      labels: rec.get("labels"),
      props: rec.get("props"),
    }));
    if (rows.length === 0) return;
    yield rows;
    lastId = rows[rows.length - 1].id;
    if (rows.length < batch) return;
  }
}

async function replicateNodes(srcSess, dstSess, labels, filter, batch, dryRun) {
  console.log("\n[nodes]");
  let total = 0;
  for (const label of labels) {
    if (filter.size > 0 && !filter.has(label)) continue;
    if (label.startsWith("_")) continue;
    let moved = 0;
    for await (const chunk of streamNodes(srcSess, label, batch)) {
      if (!dryRun) {
        await dstSess.run(NODE_UPSERT_CYPHER, { rows: chunk });
      }
      moved += chunk.length;
      total += chunk.length;
      console.log(`  · :${label}: ${moved}`);
    }
  }
  console.log(`  total nodes upserted: ${total}`);
}

// ---- Edges ----------------------------------------------------------------

async function fetchRelTypes(sess) {
  const rs = await fetchAll(
    sess,
    "CALL db.relationshipTypes() YIELD relationshipType RETURN relationshipType",
  );
  return rs.map((r) => r.relationshipType);
}

async function* streamEdges(sess, relType, batch) {
  let lastId = null;
  for (;;) {
    let res;
    if (lastId === null) {
      res = await sess.run(
        `MATCH (a)-[r:\`${relType}\`]->(b) RETURN elementId(r) AS id, elementId(a) AS startId, elementId(b) AS endId, properties(r) AS props ORDER BY elementId(r) LIMIT $limit`,
        { limit: neo4j.int(batch) },
      );
    } else {
      res = await sess.run(
        `MATCH (a)-[r:\`${relType}\`]->(b) WHERE elementId(r) > $cursor RETURN elementId(r) AS id, elementId(a) AS startId, elementId(b) AS endId, properties(r) AS props ORDER BY elementId(r) LIMIT $limit`,
        { cursor: lastId, limit: neo4j.int(batch) },
      );
    }
    const rows = res.records.map((rec) => ({
      id: rec.get("id"),
      startId: rec.get("startId"),
      endId: rec.get("endId"),
      props: rec.get("props"),
    }));
    if (rows.length === 0) return;
    yield rows;
    lastId = rows[rows.length - 1].id;
    if (rows.length < batch) return;
  }
}

const edgeCreateCypher = (t) =>
  `UNWIND $rows AS row MATCH (a:_Migrated {_neo4j_id: row.startId}) MATCH (b:_Migrated {_neo4j_id: row.endId}) CREATE (a)-[:\`${t}\` {_neo4j_id: row.id}]->(b)`;

const edgePropsCypher = (t) =>
  `UNWIND $rows AS row MATCH (a:_Migrated {_neo4j_id: row.startId})-[r:\`${t}\` {_neo4j_id: row.id}]->(b:_Migrated {_neo4j_id: row.endId}) SET r += row.props`;

async function replicateEdges(srcSess, dstSess, types, filter, batch, dryRun) {
  console.log("\n[edges]");
  let total = 0;
  for (const t of types) {
    if (filter.size > 0 && !filter.has(t)) continue;
    let moved = 0;
    const createQ = edgeCreateCypher(t);
    const propsQ = edgePropsCypher(t);
    for await (const chunk of streamEdges(srcSess, t, batch)) {
      if (!dryRun) {
        await dstSess.run(createQ, { rows: chunk });
        const hasProps = chunk.some(
          (r) => r.props && Object.keys(r.props).length > 0,
        );
        if (hasProps) {
          await dstSess.run(propsQ, { rows: chunk });
        }
      }
      moved += chunk.length;
      total += chunk.length;
      console.log(`  · :${t}: ${moved}`);
    }
  }
  console.log(`  total edges created: ${total}`);
}

// ---- Main -----------------------------------------------------------------

async function main() {
  const args = parseArgs(process.argv);
  const labelFilter = csvSet(args.labels);
  const typeFilter = csvSet(args.types);

  const srcDrv = neo4j.driver(
    args.sourceUrl,
    neo4j.auth.basic(args.sourceUser, args.sourcePass),
  );
  const dstDrv = neo4j.driver(
    args.targetUrl,
    neo4j.auth.basic(args.targetUser, args.targetPass),
  );

  const srcSess = srcDrv.session({
    database: args.sourceDb,
    defaultAccessMode: neo4j.session.READ,
  });
  const dstSess = dstDrv.session({
    database: args.targetDb,
    defaultAccessMode: neo4j.session.WRITE,
  });

  const started = Date.now();
  try {
    if (!args.skipSchema) await replicateSchema(srcSess, dstSess, args.dryRun);
    if (!args.skipNodes) {
      const labels = await fetchLabels(srcSess);
      await replicateNodes(srcSess, dstSess, labels, labelFilter, args.batchSize, args.dryRun);
    }
    if (!args.skipEdges) {
      const types = await fetchRelTypes(srcSess);
      await replicateEdges(srcSess, dstSess, types, typeFilter, args.batchSize, args.dryRun);
    }
  } finally {
    await srcSess.close();
    await dstSess.close();
    await srcDrv.close();
    await dstDrv.close();
  }

  console.log(`\nDone in ${((Date.now() - started) / 1000).toFixed(1)}s`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
